cmd/snap/cmd_routine_console_conf.go

```diff
@@ -0,0 +1,144 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/client"
+
+	"github.com/jessevdk/go-flags"
+	"github.com/snapcore/snapd/i18n"
+)
+
+type cmdRoutineConsoleConfStart struct {
+	clientMixin
+}
+
+var shortRoutineConsoleConfStartHelp = i18n.G("Start console-conf snapd routine")
+var longRoutineConsoleConfStartHelp = i18n.G(`
+The console-conf-start command starts synchronization with console-conf
+
+This command is used by console-conf when it starts up. It delays refreshes if
+there are none currently ongoing, and exits with a specific error code if there
+are ongoing refreshes which console-conf should wait for before prompting the 
+user to begin configuring the device.
+`)
+
+// TODO: move these to their own package for unified time constants for how
+// often or long we do things like waiting for a reboot, etc. ?
+var snapdAPIInterval = 2 * time.Second
+var snapdWaitForFullSystemReboot = 10 * time.Minute
+
+func init() {
+	c := addRoutineCommand("console-conf-start", shortRoutineConsoleConfStartHelp, longRoutineConsoleConfStartHelp, func() flags.Commander {
+		return &cmdRoutineConsoleConfStart{}
+	}, nil, nil)
+	c.hidden = true
+}
+
+func printfFunc(msg string, format ...interface{}) func() {
+	return func() {
+		fmt.Fprintf(Stderr, msg, format...)
+	}
+}
+
+func (x *cmdRoutineConsoleConfStart) Execute(args []string) error {
+	snapdReloadMsgOnce := sync.Once{}
+	systemReloadMsgOnce := sync.Once{}
+	snapRefreshMsgOnce := sync.Once{}
+
+	for {
+		chgs, snaps, err := x.client.InternalConsoleConfStart()
+		if err != nil {
+			// snapd may be under maintenance right now, either for base/kernel
+			// snap refreshes which result in a reboot, or for snapd itself
+			// which just results in a restart of the daemon
+			maybeMaintErr := x.client.Maintenance()
+			if maybeMaintErr == nil {
+				// not a maintenance error, give up
+				return err
+			}
+
+			maintErr, ok := maybeMaintErr.(*client.Error)
+			if !ok {
+				// if cli.Maintenance() didn't return a client.Error we have very weird
+				// problems
+				return fmt.Errorf("internal error: client.Maintenance() didn't return a client.Error")
+			}
+
+			if maintErr.Kind == client.ErrorKindDaemonRestart {
+				// then we need to wait for snapd to restart, so keep trying
+				// the console-conf-start endpoint until it works
+				snapdReloadMsgOnce.Do(printfFunc("Snapd is reloading, please wait...\n"))
```

Comment:
We probably want to add i18n.G() around this - but also probably in a followup, at least the more complicated ngettext based stuff to avoid blocking further work